Create agreement between argument type and format string

> Please correct before 2023-12-10 to safely retain your package on CRAN.

The relevant warning can be seen in GHA:

```
── R CMD check ─────────────────────────────────────────────────────────────────
* using log directory ‘/home/runner/work/vroom/vroom/check/vroom.Rcheck’
* using R Under development (unstable) (2023-11-26 r8[56](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:58)38)
* using platform: x86_64-pc-linux-gnu
* R was compiled by
    gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
    GNU Fortran (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
* running under: Ubuntu 22.04.3 LTS
* using session charset: UTF-8
* using options ‘--no-manual --as-cran’
* checking for file ‘vroom/DESCRIPTION’ ... OK
* this is package ‘vroom’ version ‘1.6.4.9000’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘vroom’ can be installed ... [61s/[59](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:61)s] WARNING
Found the following significant warnings:
  vroom_big_int.h:59:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_chr.h:41:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_dttm.h:[68](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:70):27: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_date.h:50:27: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_dbl.h:40:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_fct.h:125:29: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_int.h:38:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_num.h:50:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_time.h:50:27: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_rle.h:43:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_write.cc:2[73](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:75):16: warning: field precision specifier ‘.*’ expects argument of type ‘int’, but argument 2 has type ‘std::vector<char>::size_type’ {aka ‘long unsigned int’} [-Wformat=]
See ‘/home/runner/work/vroom/vroom/check/vroom.Rcheck/00install.out’ for details.
* used C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
* used C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
```

--- a/.github/workflows/R-CMD-check.yaml
+++ b/.github/workflows/R-CMD-check.yaml
@@ -22,6 +22,10 @@ jobs:
       fail-fast: false
       matrix:
         config:
+          # temporary additions re: vetting #524
+          - {os: windows-latest, r: 'devel'}
+          - {os: windows-latest, r: '4.2'}
+
           - {os: macos-latest,   r: 'release'}
 
           - {os: windows-latest, r: 'release'}

--- a/.github/workflows/R-CMD-check.yaml
+++ b/.github/workflows/R-CMD-check.yaml
@@ -22,10 +22,6 @@ jobs:
       fail-fast: false
       matrix:
         config:
-          # temporary additions re: vetting #524
-          - {os: windows-latest, r: 'devel'}
-          - {os: windows-latest, r: '4.2'}
-
           - {os: macos-latest,   r: 'release'}
 
           - {os: windows-latest, r: 'release'}
@@ -33,6 +29,8 @@ jobs:
           - {os: windows-latest, r: '3.6'}
           # use 4.1 to check with rtools40's older compiler
           - {os: windows-latest, r: '4.1'}
+          # keep this until the format specifier dust is truly settled (#524)
+          - {os: windows-latest, r: 'devel'}
 
           - {os: ubuntu-latest,   r: 'devel', http-user-agent: 'release'}
           - {os: ubuntu-latest,   r: 'release'}

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,5 +1,7 @@
 # vroom (development version)
 
+* Internal changes requested by CRAN around format specification (#524).
+
 # vroom 1.6.4
 
 * It is now possible (again?) to read from a list of connections (@bairdj, #514).

--- a/src/r_utils.h
+++ b/src/r_utils.h
@@ -9,14 +9,12 @@
 #include <string>
 #include <vector>
 
-#ifdef R_VERSION
-#  if (R_VERSION >= R_Version(4, 4, 0))
+#ifndef R_PRIdXLEN_T
+#  ifdef LONG_VECTOR_SUPPORT
 #    define R_PRIdXLEN_T "td"
 #  else
 #    define R_PRIdXLEN_T "d"
 #  endif
-#else
-#  define R_PRIdXLEN_T "d"
 #endif
 
 namespace vroom {

--- a/src/r_utils.h
+++ b/src/r_utils.h
@@ -9,6 +9,16 @@
 #include <string>
 #include <vector>
 
+#ifdef R_VERSION
+#  if (R_VERSION >= R_Version(4, 4, 0))
+#    define R_PRIdXLEN_T "td"
+#  else
+#    define R_PRIdXLEN_T "d"
+#  endif
+#else
+#  define R_PRIdXLEN_T "d"
+#endif
+
 namespace vroom {
 
 inline std::string

--- a/src/unicode_fopen.h
+++ b/src/unicode_fopen.h
@@ -49,7 +49,7 @@ inline FILE* unicode_fopen(const char* path, const char* mode) {
   }
   buf = (wchar_t*)R_alloc(len, sizeof(wchar_t));
   if (buf == NULL) {
-    Rf_error("Could not allocate buffer of size: %ll", len);
+    Rf_error("Could not allocate buffer of size: %zu", len);
   }
 
   MultiByteToWideChar(CP_UTF8, 0, path, -1, buf, len);
@@ -75,7 +75,7 @@ make_mmap_source(const char* file, std::error_code& error) {
   }
   buf = (wchar_t*)malloc(len * sizeof(wchar_t));
   if (buf == NULL) {
-    Rf_error("Could not allocate buffer of size: %ll", len);
+    Rf_error("Could not allocate buffer of size: %zu", len);
   }
 
   MultiByteToWideChar(CP_UTF8, 0, file, -1, buf, len);

--- a/src/utils.h
+++ b/src/utils.h
@@ -5,6 +5,7 @@
 #include <cstring>
 #include <sstream>
 #include <string>
+#include <tuple>
 
 namespace vroom {
 

--- a/src/vroom_big_int.h
+++ b/src/vroom_big_int.h
@@ -6,8 +6,8 @@
 
 constexpr long long NA_INTEGER64 = 0x8000000000000000LL;
 
-#include "vroom.h"
 #include "r_utils.h"
+#include "vroom.h"
 
 namespace cpp11 {
 inline bool is_na(long long x) { return x == NA_INTEGER64; }

--- a/src/vroom_big_int.h
+++ b/src/vroom_big_int.h
@@ -7,6 +7,7 @@
 constexpr long long NA_INTEGER64 = 0x8000000000000000LL;
 
 #include "vroom.h"
+#include "r_utils.h"
 
 namespace cpp11 {
 inline bool is_na(long long x) { return x == NA_INTEGER64; }
@@ -56,9 +57,10 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_big_int (len=%d, materialized=%s)\n",
-        Length(x),
-        R_altrep_data2(x) != R_NilValue ? "T" : "F");
+      "vroom_big_int (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
+      Length(x),
+      R_altrep_data2(x) != R_NilValue ? "T" : "F");
+
     return TRUE;
   }
 

--- a/src/vroom_big_int.h
+++ b/src/vroom_big_int.h
@@ -57,10 +57,9 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-      "vroom_big_int (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
-      Length(x),
-      R_altrep_data2(x) != R_NilValue ? "T" : "F");
-
+        "vroom_big_int (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
+        Length(x),
+        R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;
   }
 

--- a/src/vroom_chr.h
+++ b/src/vroom_chr.h
@@ -39,10 +39,9 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-      "vroom_chr (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
-      Length(x),
-      R_altrep_data2(x) != R_NilValue ? "T" : "F");
-
+        "vroom_chr (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
+        Length(x),
+        R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;
   }
 

--- a/src/vroom_chr.h
+++ b/src/vroom_chr.h
@@ -4,6 +4,7 @@
 
 #include "altrep.h"
 
+#include "r_utils.h"
 #include "vroom_vec.h"
 
 cpp11::strings read_chr(vroom_vec_info* info);
@@ -38,9 +39,10 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_chr (len=%d, materialized=%s)\n",
-        Length(x),
-        R_altrep_data2(x) != R_NilValue ? "T" : "F");
+      "vroom_chr (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
+      Length(x),
+      R_altrep_data2(x) != R_NilValue ? "T" : "F");
+
     return TRUE;
   }
 

--- a/src/vroom_date.h
+++ b/src/vroom_date.h
@@ -2,6 +2,7 @@
 
 #include <cpp11/doubles.hpp>
 
+#include "r_utils.h"
 #include "vroom_dttm.h"
 
 using namespace vroom;
@@ -47,7 +48,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_date (len=%d, materialized=%s)\n",
+        "vroom_date (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_dbl.h
+++ b/src/vroom_dbl.h
@@ -2,6 +2,7 @@
 
 #include "altrep.h"
 #include "parallel.h"
+#include "r_utils.h"
 #include "vroom_vec.h"
 
 double bsd_strtod(const char* begin, const char* end, const char decimalMark);
@@ -37,7 +38,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_dbl (len=%d, materialized=%s)\n",
+        "vroom_dbl (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_dttm.h
+++ b/src/vroom_dttm.h
@@ -3,6 +3,7 @@
 #include <cpp11/doubles.hpp>
 #include <cpp11/integers.hpp>
 
+#include "r_utils.h"
 #include "vroom.h"
 #include "vroom_vec.h"
 
@@ -65,7 +66,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_dttm (len=%d, materialized=%s)\n",
+        "vroom_dttm (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_fct.h
+++ b/src/vroom_fct.h
@@ -3,6 +3,7 @@
 #include <cpp11/strings.hpp>
 
 #include "altrep.h"
+#include "r_utils.h"
 #include "vroom.h"
 #include "vroom_vec.h"
 
@@ -122,7 +123,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_factor (len=%d, materialized=%s)\n",
+        "vroom_factor (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_int.h
+++ b/src/vroom_int.h
@@ -2,6 +2,7 @@
 
 #include "altrep.h"
 
+#include "r_utils.h"
 #include "vroom_vec.h"
 
 int strtoi(const char* begin, const char* end);
@@ -35,7 +36,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_int (len=%d, materialized=%s)\n",
+        "vroom_int (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_num.h
+++ b/src/vroom_num.h
@@ -4,6 +4,7 @@
 
 #include "altrep.h"
 
+#include "r_utils.h"
 #include "vroom_vec.h"
 
 #include "parallel.h"
@@ -47,7 +48,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_num (len=%d, materialized=%s)\n",
+        "vroom_num (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_rle.h
+++ b/src/vroom_rle.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "altrep.h"
+#include "r_utils.h"
 
 #ifdef HAS_ALTREP
 
@@ -40,7 +41,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_rle (len=%d, materialized=%s)\n",
+        "vroom_rle (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_time.h
+++ b/src/vroom_time.h
@@ -2,6 +2,7 @@
 
 #include <cpp11/doubles.hpp>
 
+#include "r_utils.h"
 #include "vroom.h"
 #include "vroom_dttm.h"
 
@@ -47,7 +48,7 @@ public:
   static Rboolean
   Inspect(SEXP x, int, int, int, void (*)(SEXP, int, int, int)) {
     Rprintf(
-        "vroom_time (len=%d, materialized=%s)\n",
+        "vroom_time (len=%" R_PRIdXLEN_T ", materialized=%s)\n",
         Length(x),
         R_altrep_data2(x) != R_NilValue ? "T" : "F");
     return TRUE;

--- a/src/vroom_write.cc
+++ b/src/vroom_write.cc
@@ -270,7 +270,7 @@ void write_buf_con(const std::vector<char>& buf, SEXP con, bool is_stdout) {
   if (is_stdout) {
     std::string out;
     std::copy(buf.begin(), buf.end(), std::back_inserter(out));
-    Rprintf("%.*s", buf.size(), out.c_str());
+    Rprintf("%.*s", (int) buf.size(), out.c_str());
   } else {
     write_buf(buf, con);
   }

